randovania/game_description/hint.py

```diff
@@ -122,33 +112,71 @@ def from_json(cls, json_dict: dict, **extra: typing.Any) -> typing.Self:
 
 @dataclass(frozen=True)
 class Hint(JsonDataclass):
-    # FIXME: refactor this into subclasses. this is a typing nightmare
-    hint_type: HintType
+    @classmethod
+    def from_json(cls, json_dict: dict, **extra: typing.Any) -> Hint:
+        hint_type = HintType(json_dict["hint_type"])
+
+        return hint_type.hint_class.from_json(json_dict, **extra)
+
+    @property
+    def as_json(self) -> dict:
+        data = super().as_json
+        data["hint_type"] = {v: k for k, v in HINT_TYPE_TO_CLASS.items()}[self.__class__].value
```

Comment:
Couldn't `Hint` have a `classmethod` or something that each one implements that returns the `HintType`?